powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/CallbackUrlBehavior.java

```diff
@@ -154,8 +154,8 @@ public UpdateCallbackUrlResponse updateCallbackUrl(UpdateCallbackUrlRequest requ
         }
 
         final Optional<CallbackUrlEntity> entityOptional = callbackUrlRepository.findById(request.getId());
-        if (entityOptional.isEmpty()) {
-            logger.warn("Invalid callback ID: "+request.getId());
+        if (entityOptional.isEmpty() || !request.getApplicationId().equals(entityOptional.get().getApplication().getId())) {
+            logger.warn("Invalid callback ID: {}", request.getId());
             // Rollback is not required, error occurs before writing to database
             throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
         }
```

Comment:
First of all, I would like to see a unit test for that.
FYI, it may be rewritten to Java 8 style, for example
```suggestion
        final CallbackUrlEntity entity = entityOptional
                .filter(it -> it.getApplication().getId().equals(request.getApplicationId()))
                .orElseThrow(() -> {
                    // Rollback is not required, error occurs before writing to database
                    logger.warn("Invalid callback ID: {}", request.getId());
                    return localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
                });
```